GEP-1897: TLS from Gateway to Backend for ingress

candita opened the pull request:
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
There is a well-known gap in the API specification regarding TLS from Gateway to Backend for ingress.

This GEP is _very_ tightly scoped because we have tried and failed to address this well-known gap in the API specification. The lack of support for this fundamental concept is holding back Gateway API adoption by users that require a solution to the use case. 

**Which issue(s) this PR fixes**:

Fixes #1897

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

bowei commented:
/ok-to-test

bowei commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,113 @@
+
+## Goals
+
+1. The solution should satisfy the following use case: the backend pod has its own
```
We should mention the approach to the x-ns reference between Route and Service.

As mentioned on the call -- it would help to say explicitly that we are going to use the standard rules wrt to direct reference configuration.

candita replied:
Thanks @bowei, I will be sure to mention this in the implementation section, as I think the description will be dependent on the implmentation details.

robscott replied:
Agree that a `TODO` and/or Unresolved section about this so we don't lose track of it. Let me know if this already happened and I missed it.

candita replied:
I will add an Open Questions section and try to record anything having to do with implementation details.

candita replied:
Added to Open Questions.

hzxuzhonghu commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,113 @@
+
+Given that the current ingress solution specifies **edge** TLS termination (from the edge to
+the gateway), and how to handle **passthrough** TLS (from the edge to the backend pod), this
+proposed ingress solution specifies **backend** TLS termination (from the gateway to the
```
IIRC, only simple tls origination is covered here and mTLS is not

bowei replied:
FWIW -- I do think this is important to call out -- the service is not in scope for the GW API.

shaneutt replied:
I have a suggestion which I think may help to resolve the concerns here:

```suggestion
proposed ingress solution specifies TLS origination to the **backend** (from the gateway to the
```

The comment in parenthesis might need a bit more tweaking.

shaneutt reviewed (approved):
I think this is a great example of a "what and why without the how" initial GEP to start back up the conversation on this topic without having too many details to comb through right away. As a maintainer I particularly appreciate the iterative approach in the face of our [focus on GA](https://docs.google.com/document/d/1GcPK4cCkbyW6iGwSmBk6Q1dOSA_hGtq02aUx7TwjlTw) as it helps to spread the cognitive load over smaller bits of time, and makes it more feasible for me to review in between other priorities.

The suggestion that this is an important feature for multiple implementations rings true with me, the goals and non-goals make sense, and I very much appreciate you meticulously laying out the long and troubled history of this effort's past.

As I have no blockers and this is `Provisional` with more iterations to come, I approve and I'm looking forward to seeing the follow-ups.

As has become customary with GEPs recently I'll place a temporary hold so that we can allow some time for several people to review without it merging too fast.

/hold

Thanks @candita!

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,113 @@
+1. Changes to the existing mechanisms for edge or passthrough TLS termination
+2. Providing a mechanism to decorate multiple route instances
+3. Service mesh use cases
+4. Mutual TLS use cases
+5. TLSRoute, TCPRoute, or UDPRoute use cases
```
While I acknowledge that it is critical to keep the scope reduced in order to make forward progress, I am a bit concerned with making these "Non-goals" (these 2 in particular, but to some extent egress and mesh as well).

I worry that by only considering the narrowest case, we will pigeon-hole ourselves into a non-flexible API. I don't want us to implement "TLSSettingsForHTTPFromIngress" (logically or literally) now, and realize we actually wanted it to be a more general "TLSSettingsFromIngress" (support TCPRoute) or "TLSSettings" (support non-ingress) cases and not be able to move.

Its possible that we will just get lucky and this won't happen, but relying on that seems unlikely to work out well.

I would feel more comfortable if the GEP was more like https://github.com/kubernetes-sigs/gateway-api/pull/1643, which I view as the golden standard for a "What and why" GEP, which explores the problem space and prior art.

Without this, I worry we are just going to solve one very very narrow case and create an awkward long term API.

I would also like to remind folks that the easiest way to experiment with these type of customizations is to _just go implement it_. Gateway API is designed around extensibility; there is nothing stopping a custom implementation from implementing TLS origination today. This serves as an incredibly valuable reference point for discussion - at the end of the day, we get far more info from a real implementation than debating on a doc. That isn't to say that this isn't valuable in core - it absolutely is. But its important that we do it right in core than that we do it fast (or even at all!).

shaneutt replied:
As this comment stands there's a lot of room for interpretation and there's some ambiguity in the suggestions for action. In this state I could see it hanging around for some time without being resolved, so I want to try and get it distilled. The "just go implement it" suggestion is not something we the maintainers feel we should be pushing for in this case, but for everything else we would recommend creating prescriptive and explicit suggestions (ideally with Github's review suggestions feature when feasible, as these are clear and actionable). Are you OK with resolving this comment in favor of reworking it into more explicit suggestions for change?

costinm commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,113 @@
+2. In terms of the Gateway API personas, the application developer should control the
+gateway to backend TLS settings, not the cluster operator, as requiring a cluster operator
+to manage certificate renewals and revocations would be extremely cumbersome.
+3. The solution should consider client certificate settings used in the TLS handshake **from
```
I don't think TLS versions or cipher suites are appropriate - should be out of scope as well. 

They are typically restricted for compliance reasons  - an app owner should not be able to override. Also platform owner may enforce this by having infastructure restrict this ( compile only the compliant subset, etc). 

Another reason is that they require expertise that app owners typically don't have - very few people understand the compliance space.

candita replied:
They are in scope for this discussion, as TLS version and cipher suites can be application dependent.  Whether the app owner can override the platform owner is an implementation detail too.

howardjohn replied:
My understanding:
* Apps will set allowed version/suites - this is not something we decide, that is how things work.
* Per costin, platform owners _should_ have a way to enforce gateway-api driven defaults.
* TLS version/cipher is a handshake

Where we may want a per-app/per-service gateway-api config override is if the platform admin says (for example) "TLS 1.3 is required", but an app says "I only speak TLS 1.2". As is, that would fail. However, we could allow a per-service opt-out of the platform rules and say "for only svc-legacy, allow 1.2".

This seems a bit strange to me, though. If we want TLS 1.3 everywhere, we probably _really_ want it everywhere. Its not clear why we would be ok with less secure methods for one service but not another.

If we just want to default to more secure options, that is very different - as mentioned, TLS is a handshake negotiation on these, so we can say at admin level "Support 1.3,1.2 and suites [best,2ndbest,3rdbest]" or equivilent.

candita replied:
I agree that an override or default should be possible.  However, I will leave this discussion for another GEP or the implementation details.  Every cluster has its use cases, and we may need to address a cluster-level override or default with a different set of use cases.  Sound fair?

howardjohn replied:
I don't think this is an "overrides and defaults" scenario. We need that when we are talking about something like Retry or Timeout. TLS is different; it is an existing protocol which already has a negotiation aspect. We cannot and should not invent another negotiation on top of that.

costinm replied:
+1

Also UX is important - FIPS compliance and what cipher suites are safe or required is a completely different persona than Jane the app developer.  It is something that may be implemented by compile options and deep into platform - and if a cipher or TLS version becomes unsafe it is not up to the developers to make changes, but platform owner.

youngnick replied:
@howardjohn and @costinm, everything you've said is correct, but you are missing the function that end users end up using an ingress proxy for - fixing broken apps. It's _very_ common that the TLS version policy will be "Everything must be TLS 1.3, except for this list of exceptions". Having a way to describe that list of exceptions is imperative.

Yes, this can absolutely be a footgun, but it is _definitely_ required in real-world use cases. We had to add similar functionality to Contour for exactly this reason.

Does this create effectively another negotiation over the top of the TLS one? Yes. Is it still unfortunately necessary? Also yes.

This is just one of those things that people who are stuck working in a service-mesh free world are stuck with.

shaneutt replied:
Considering this resolved as there hasn't been an update in some time. If you feel there's still action that needs to be taken here please unresolve and let us know in a comment.

costinm commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,113 @@
+backend pod. (Use case #4 in [Gateway API TLS Use Cases](#references))
+2. In terms of the Gateway API personas, the application developer should control the
+gateway to backend TLS settings, not the cluster operator, as requiring a cluster operator
+to manage certificate renewals and revocations would be extremely cumbersome.
```
On client certificate - I don't think an API is appropriate. App code can ask for them ( using TLS protocol ), the API can't influence this. And the Gateway - if it can - should present its identity. App owner can't choose a random client cert.

costinm replied:
Also - Gateway SHOULD authenticate with either a JWT with audience or client cert or some other mean - so gateway added headers can be trusted, etc.

candita replied:
I am referring to the application developer as a Gateway API persona, and thus the person who creates the HTTPRoute.  The second comment, iiuc is implementation dependent.

costinm commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,113 @@
+covers too much ground to address in a single GEP.
+
+## Prior Art - the history of backend TLS
+
```
I think Prior Art is about existing implementations - like Istio DR, Ingress annotations, APIs supported by various load balancers. 

At least that's what we do for persistent sessions - previous discussions are not prior art.

shaneutt replied:
It seems fine to me to change the language, Github suggestions can be really helpful for this because it shows the contributor a clear and prescriptive action they can take to consider your comment resolved, as such I created this one:

https://github.com/kubernetes-sigs/gateway-api/pull/1906/files#r1157838147

(I had to create a new comment rather than commenting here because you targeted line 60 which was the line below)

Let me know if you're fine with how that's worded, or want something different, or feel free to make a suggestion in the same thread and I'll rescind mine.

I'll consider this resolved so we can follow up over there.

costinm reviewed (commented):
I think it's a valid use case and it is ok to start with a narrow set of use cases ( as long as we believe the API choice wouldn't prevent or make difficult the other use cases). 

However I would requiest that "Prior art" section is fixed and we include the actual 
APIs in current use by different LB solutions. Past internal discussions are not prior art.
Once we have a clear picture of what various implementations currently do and support - we will have a much easier time defining our API.

shaneutt commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,113 @@
+is that we have been too focused on a generic representation of everything that TLS can do, which
+covers too much ground to address in a single GEP.
+
+## Prior Art - the history of backend TLS
```
Based on feedback from @costinm in https://github.com/kubernetes-sigs/gateway-api/pull/1906/files#r1157820085

```suggestion
## The history of backend TLS
```

liwenwu-amazon commented:
The GEP looks good!. In [AWS VPC Lattice controller gateway API/multicluster implementation](https://github.com/aws/aws-application-networking-k8s),  VPC Lattice is the federation.   

For serviceexport, the lattice controller register all of its endpoints in local cluster to VPC lattice

To reference a serviceimport as a backend of a HTTPRoute, the lattice controller will configure VPC lattice resource accordingly. So that the traffic will get routed to endpoints in that remote cluster(VPC) by VPC lattice dataplane.

In addition to K8S endpoints, serviceimport can references targets like EC2, lambda function and ECS container(s).

candita commented:
@howardjohn @costinm @bowei I have highlighted some goals as immediate and longer term, as well as added some details on prior art in several implementations. PTAL when you get a chance.

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,172 @@
+solution. The application developer should control the gateway to backend TLS settings,
+not the cluster operator, as requiring a cluster operator to manage certificate renewals
+and revocations would be extremely cumbersome.
+3. The solution should consider client certificate settings used in the TLS handshake **from
```
sorry to nitpick but version and cipher suites are TLS handshake properties, not "certificate settings"

candita replied:
Fixed

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,172 @@
+2. Providing a mechanism to decorate multiple route instances
+3. TLSRoute use cases
+4. UDPRoute use cases 
+5. Termination of TLS for HTTP routing (#1 in [Gateway API TLS Use Cases](#references))
+6. HTTPS passthrough use cases (#2 in [Gateway API TLS Use Cases](#references))
+7. Termination of TLS for non-HTTP TCP streams (#3 in [Gateway API TLS Use Cases](#references))
```
Mentioning these as "already solved" may be more clear than "Non-goals".

candita replied:
I prefer to cover all the use cases in the referenced TLS use cases document.  Sound fair?

robscott replied:
Agree with @howardjohn here that it would be better to distinguish between "already solved" and "not yet done but not in scope".

candita replied:
I will add a section "Already Solved TLS Use Cases" to isolate these 3 cases.

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,172 @@
+6. HTTPS passthrough use cases (#2 in [Gateway API TLS Use Cases](#references))
+7. Termination of TLS for non-HTTP TCP streams (#3 in [Gateway API TLS Use Cases](#references))
+8. Controlling certificates used by more than one workload (#6 in [Gateway API TLS Use Cases](#references))
+9. Client certificate settings used in the TLS handshake **from external clients to the
```
I don't understand this. Its not possible for us to control and external client?

I get its a Non-Goal; but checking my understanding here

candita replied:
This is [use case 7](https://docs.google.com/document/d/17sctu2uMJtHmJTGtBi_awGB0YzoCLodtR6rUNmKMCs8/edit#heading=h.y0ofkfsm5t0c) , which @youngnick mentioned is "particularly useful for JWT use cases, and can make doing some types of authentication and authorization easier".

candita replied:
Removed the word "handshake".

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,172 @@
+8. Controlling certificates used by more than one workload (#6 in [Gateway API TLS Use Cases](#references))
+9. Client certificate settings used in the TLS handshake **from external clients to the
+Listener** (#7 in [Gateway API TLS Use Cases](#references))
+10. The cluster operator should be allowed to override any gateway to backend TLS settings, to
```
Its a bit confusing to say "Should be allowed to" in "non-goals".

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,172 @@
+backend pod).  As mentioned, this solution satisfies the use case in which the backend pod
+has its own certificate and the gateway client needs to know how to connect to the backend pod.
+
+![image depicting TLS termination types](images/1897-TLStermtypes.png "TLS termination types")
```
very nit, would be nice to have one showing terminating but not originating.

candita replied:
Hm, this is the exact opposite of the first feedback I got on this diagram.

candita replied:
I originally had the three termination types separated in a diagram, and got feedback from @robscott that it made more sense to combine these two in one as a contrast to the passthrough route. I can't please both of you, but hope it's okay as is.

robscott replied:
Yeah for a bit more context it was very difficult to show all possible forms here. We'd essentially need to cover all of the following options:

1. Gateway termination, no origination
2. Gateway termination, origination
3. No Gateway termination, origination
4. No Gateway termination, no origination
5. Passthrough

The diagram that we ended up with feels simpler than anything that tried to represent all possible cases, and representing a slightly larger subset could just lead to confusion.

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,172 @@
+
+### Istio Gateway supports this with a DestinationRule to encrypt the gateway->backend hop:
+
+* A secret representing a certificate/key pair, where the certificate is valid for the route host
+* Set Gateway spec.servers[].port.protocol: HTTPS, spec.servers[].tls.mode=SIMPLE, spec.servers[].tls.credentialName:
```
nit: this is technically not "encrypt the gateway->backend", this is terminating client->gateway.

Only `Set DestinationRule spec.trafficPolicy.tls.mode: SIMPLE` is "encrypt the gateway->backend"

candita replied:
Updated

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,172 @@
+### GKE supports a way to encrypt traffic to the backend pods using:
+
+* `AppProtocol` on Service set to HTTPS
+* Certificate verification by pod
```
Not sure what this means but I don't think GKE verifies anything

candita replied:
Updated

howardjohn commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,172 @@
+
+Details deferred until we reach consensus on what we want to do, and why we want to do this.
+
+## Prior Art
```
So:

* Client decides TLS: Istio, Ambassador, OpenShift
* Server (service) decides TLS: Contour, GKE

Is that a fair representation?

candita replied:
I think Ambassador relies on the service configuration, so I would put it with the second bullet.

costinm replied:
AFAIK Istio is both - with the new auto-mtls clearly in 'server decide' category.

And the service ultimately decides everywhere - if the service has a TLS port, there is nothing a client can do to change the fact that only TLS connections will work.

candita replied:
@howardjohn @costin If there's something you want me to add to summarize this better, or you want me to point out the two categorizations, please let me know.

costinm commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,171 @@
+## Overview - what do we want to do?
+
+Given that the current ingress solution specifies **edge** TLS termination (from the client to
+the gateway), and how to handle **passthrough** TLS (from the client to the backend pod), this
```
Note that 'passthrough' is covered by TLSRoute ( routing via SNI without edge termination ). 

Gateway communicating with a backend using TLS has nothing to do with edge TLS termination - it is just a property of the backend that it requires TLS. Even if the original internet request was plain HTTP - the gateway would still have to use TLS to the backend ( because typically that's all that the backend accepts ).

That's why 're-encryption' or 'passthrough' are not right terms here.

youngnick replied:
This section is saying that TLS termination at the _Listener_ is already done, as is TLS Passthrough (with TLSRoute). It's entirely appropriate to use those terms here in those two cases.

costinm commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,171 @@
+Gateway API is missing a mechanism for separately providing the details for the backend TLS handshake,
+including:
+* use of TLS
+* destination CA (certificate authority) or CA bundle
```
Not sure if defining K8S root CAs is really in scope for this WG - my understanding was that there are separate WG working on certificate provisioning ( cert request, etc) and root CA configurations. 

Mutating/Validating webhooks are a major pain - I would be very careful around this UX, it is extremely hard to automate or use in helm charts.

youngnick replied:
This is _not_ about defining K8S root CAs. Read the text.

This is about _being able to specify a CA bundle for the connection from the Gateway to the backend_. It's common that a service owner that is doing their own certificate management will manage their own root of trust, which will necessitate a separate CA bundle. It has nothing to do with Mutating or Validating webhooks.

shaneutt replied:
Indeed, but if you feel we've missed something and there's some changes to the language needed here please do drop a specific suggestion for change and/or let us know.

costinm commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,171 @@
+including:
+* use of TLS
+* destination CA (certificate authority) or CA bundle
+* SANs for validating upstream service (server authentication)
```
This is another area where I'm not sure the HttpRoute author should make decisions ( egress is a slightly different case that I believe is not in scope here ). 

Currently Spiffee (URL) and DNS SANs are the majority of SANs in use. There are use cases where the DNS SAN may not be the same as authority - but not clear that the HttpRoute author is the best persona for configuring this ( as opposed to the service owner ).

shaneutt replied:
Indeed, we're just trying to acknowledge what the problem is and if it's worth working on. That said, if you feel there's a way we can better capture this and other comments which aren't clearly actionable at this stage, we've used a `TODO` list inside the GEP elsewhere it may be reasonable to suggest TODO items that need to be resolved before we move out of `Provisional` for some of your concerns, do feel free to suggest adding that to this PR.

candita replied:
Added to Open Questions.

costinm commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,171 @@
+
+TLS from gateway to backend for ingress exists in several implementations, and was developed independently.
+
+### Istio Gateway supports this with a DestinationRule:
```
DestinationRule is mainly using this for egress destinations. 

For Gateway -> backend, Istio (now) uses auto-mtls, with the root CAs distributed as a config map and not explicitly configured by users.

Using tls.mode=SIMPLE and the rest for internal traffic is a rarely used config - and typically breaks Istio telemetry and policy features.

shaneutt replied:
What action would you like to see taken here?

youngnick commented on geps/gep-1897.md:
```diff
@@ -0,0 +1,171 @@
+Given that the current ingress solution specifies **edge** TLS termination (from the client to
+the gateway), and how to handle **passthrough** TLS (from the client to the backend pod), this
+proposed ingress solution specifies TLS origination to the **backend** (from the gateway to the
+backend pod).  As mentioned, this solution satisfies the use case in which the backend pod
```
Again, the point of this initial cut of this GEP is to agree that _the problem is worth solving_. Determining who owns the config and what the config looks like are part of the implementation, which will be a separate PR. This is not the time for this feedback.

shaneutt replied:
Indeed, the implementation details will come and most certainly will take a while to tease through.

candita replied:
Added to Open Questions.

youngnick commented:
/lgtm for provisional.

/hold for more reviewers though.

robscott commented:
Would like to see a couple small follow ups on comment threads (https://github.com/kubernetes-sigs/gateway-api/pull/1906#discussion_r1176961243 and https://github.com/kubernetes-sigs/gateway-api/pull/1906#discussion_r1176959862), but otherwise this LGTM. Thanks @candita!

candita commented:
@howardjohn @costinm @robscott @bowei @shaneutt I have made the following changes:

- Removed already-solved use cases from non-goals
- Added these uses cases to a new section called Already Solved TLS Use Cases
- Added a new section called Open Questions (TODO) listing comments from @bowei and @costinm

candita commented:
/label tide/merge-method-squash

robscott commented:
Will leave this for someone else to add the final LGTM, thanks for all the work on this @candita!

/approve

shaneutt reviewed (approved):
/lgtm

youngnick commented:
/unhold
